<a href="https://colab.research.google.com/github/iankipchumba06-max/machine-learning-in-finance--assignment/blob/main/Copy_of_Week1_NumPy_ML_in_Finance_Final_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<html> <h1 style="font-style:bold; color:blue;"> Machine Learning in Finance </h1> </html>

<html> <h1 style="font-style:italic; color:blue;"> Week-1 </h1> </html>


In [83]:
import  numpy as np

# NumPy

The main python library used in numerical data analysis is NumPy (or Numpy). It is a Linear Algebra Library for Python.  It is a very important library for Data Science with Python mainly because Numpy is incredibly fast, as it has bindings to C libraries.

It provides a high-performance multidimensional array object, and tools for working with these arrays.

This practical gives only an introduction to Numpy.

## Installation NumPy

**In the case of using the Anaconda distribution, the Numpay has already been installed. If you have any problem with Numpay functionality, it is highly recommended to re-install Numpay using the following command:**
    
    install numpy


____________________________________________![image.png](attachment:d7dc4e19-f52e-4a93-b2f1-e4cd231d4be9.png)

In [84]:
a =np.array([1,2,3])

In [86]:
print(a[0],a[1],[2])
print(a.shape)

1 2 [2]
(3,)


One of the most popular features of the Numpy library is the capability of operations with arrays of different sizes (or ranks).

Dimensions of arrays:
- vectors - One dimensional arrays,
- matrices - Two dimensional arrays,
- tensor - Three or more arrays.

49.png![image.png](attachment:image.png)

____________________________________________![image.png](attachment:c8b0e131-8ac3-4605-b402-9cca1b5ba9d3.png)

In [87]:
a[0] = 5

In [88]:
b = np.array([[1,2,3],[4,5,6]])

____________________________________________![image.png](attachment:565f7ffe-ea75-4f62-b130-093f60e7e041.png)

In [89]:
print(b [1, 0])
print(a.shape)

4
(3,)


In [90]:
b = np.array([[1,2,3], [4,5,6]])

### arange

The arange() method is used to return an evenly spaced values within a given interval.
The method has the following parameters:
- **start** parameter (inclusive in the sequence)  
- **end** parameter (exclusive in the sequence)
- **step size** (optional parameter used to find the next element in the sequence)


____________________________________________![image.png](attachment:be72cd71-324f-4b2f-a9eb-9f58864ee035.png)

In [91]:
np.arange(10,20)

array([10, 11, 12, 13, 14, 15, 16, 17, 18, 19])

____________________________________________![image.png](attachment:793515fa-88f9-471d-bb3d-6b28e65a20c1.png)

In [92]:
np.arange(30,1,-3)

array([30, 27, 24, 21, 18, 15, 12,  9,  6,  3])

<html> <h3 style="font-style:italic; color:blue;"> Let us compare the performance of Numpay and Python arrays. </h3> </html>

In [93]:
# timeit - executes the code several times and gives statistics.

#### Now just press 'CTRL' and 'ENTER' keys on the next cells.

In [26]:
%timeit np.arange(0,1000)

1.34 µs ± 162 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


In [27]:
%timeit range(0,1000)

107 ns ± 27.6 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)


<html> <h3 style="color:blue;"> Numpy also provides many functions to create arrays </h3> </html>

____________________________________________![image.png](attachment:5644d1a1-af6f-499a-8b57-56e2b163cf07.png)

In [28]:
a = np.zeros((2,2))

____________________________________________![image.png](attachment:2b89ec2a-047e-4e58-9ca8-fcb0f4cb6b5b.png)

In [29]:
b = np.ones((3,4))

____________________________________________![image.png](attachment:267a89d5-8030-4262-b0b8-6c744458ad89.png)

In [30]:
np.eye(4)

array([[1., 0., 0., 0.],
       [0., 1., 0., 0.],
       [0., 0., 1., 0.],
       [0., 0., 0., 1.]])

In [31]:
e =np.random.random((2,2))

____________________________________________![image.png](attachment:339d2c8c-51d3-4874-9dd6-ccc57337eeb1.png)

In [32]:
np.linspace(0,20,11)

array([ 0.,  2.,  4.,  6.,  8., 10., 12., 14., 16., 18., 20.])

In [33]:
np.linspace(0,20,30)

array([ 0.        ,  0.68965517,  1.37931034,  2.06896552,  2.75862069,
        3.44827586,  4.13793103,  4.82758621,  5.51724138,  6.20689655,
        6.89655172,  7.5862069 ,  8.27586207,  8.96551724,  9.65517241,
       10.34482759, 11.03448276, 11.72413793, 12.4137931 , 13.10344828,
       13.79310345, 14.48275862, 15.17241379, 15.86206897, 16.55172414,
       17.24137931, 17.93103448, 18.62068966, 19.31034483, 20.        ])

____________________________________________![image.png](attachment:155cb7d7-48d9-4b79-a551-503d42cca2b9.png)

<html> <h3 style="color:blue;"> Array Indexing </h3> </html>

**Slicing:**
    Similar to Python lists, numpy arrays can be sliced. Since arrays may be multidimensional, you must specify a slice for each dimension of the array:

..................................................![image.png](attachment:11cabfa7-9787-4a3a-94d5-f27b5089bbea.png)

In [34]:
a = np.array([[1,2,3,4],[5,6,7,8],[9,10,11,12]])

In [35]:
b = a[:2,1:3]

____________________________________________![image.png](attachment:b4086256-1bbb-4e1c-820c-415d37eb11b2.png)

In [36]:
print(a[0,1])

2


____________________________________________![image.png](attachment:f3b12e40-eda9-467d-9711-d9b57f20f6ea.png)

In [37]:
b[0,0] = 88
print(a)

[[ 1 88  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]]


____________________________________________![image.png](attachment:3bf651f2-ee5c-4cf3-a41c-8bbfd13bd70f.png)

<html> <h3 style="color:blue;"> Datatypes </h3> </html>

Every NumPy array is a grid of elements of the **same type**. Numpy provides a large set of numeric datatypes that you can use to construct arrays. Numpy tries to guess a datatype when you create an array, but you can explicitly specify the type using an optional argument.

____________________________________________![image.png](attachment:acf9492f-76d6-451f-90d6-f8af21190883.png)

In [38]:
x = np.array([1,2])
print(x.dtype)

int64


____________________________________________![image.png](attachment:abd5f58f-2ced-436f-bf31-f5455975721d.png)

In [39]:
x = np.array([1.0,2.0])
print(x.dtype)

float64


____________________________________________![image.png](attachment:cc855679-b5ab-4be8-859f-cbecf8a0fbba.png)

In [40]:
x = np.array([1,2], dtype=np.int64)
print(x.dtype)

int64


<html> <h3 style="color:blue;"> Array Math </h3> </html>

In [41]:
# Basic mathematical functions operate elementwise on arrays

.....................................................![image.png](attachment:62b9a0d9-3d99-4b14-a8ad-3711111afd5f.png)

#### Now just press 'CTRL' and 'ENTER' keys on the next cell.

In [42]:
x = np.array([[1,2],[3,4]], dtype=np.float64)
y = np.array([[5,6],[7,8]], dtype=np.float64)
print(x, '\n')   #'\n'
print(y)

[[1. 2.]
 [3. 4.]] 

[[5. 6.]
 [7. 8.]]


____________________________________________![image.png](attachment:c0c48780-516e-48fc-93be-796adc25d912.png)

In [43]:
print(x + y)
print('\n')
print(np.add(x, y))

[[ 6.  8.]
 [10. 12.]]


[[ 6.  8.]
 [10. 12.]]


In [44]:
print(x -y,  '\n')
print(np.subtract(x,y))

[[-4. -4.]
 [-4. -4.]] 

[[-4. -4.]
 [-4. -4.]]


In [45]:
print(x * y, '\n')
print(np.multiply(x,y))


[[ 5. 12.]
 [21. 32.]] 

[[ 5. 12.]
 [21. 32.]]


____________________________________________![image.png](attachment:b6a795f3-e338-490f-a256-39a511082445.png)

In [46]:
print(x / y)
print('\n')
print(np.divide(x, y))

[[0.2        0.33333333]
 [0.42857143 0.5       ]]


[[0.2        0.33333333]
 [0.42857143 0.5       ]]


____________________________________________![image.png](attachment:07714d60-442c-4abd-b584-a8d7af18d9bb.png)

In [47]:
print(np.sqrt(x))

[[1.         1.41421356]
 [1.73205081 2.        ]]


____________________________________________![image.png](attachment:3ad4361b-8759-4cf4-acda-29ef1b7d6ffc.png)

In [48]:
x**3

array([[ 1.,  8.],
       [27., 64.]])

____________________________________________![image.png](attachment:a6875e9c-4f66-42f3-a098-5d810431a65c.png)

In [49]:
1/x

array([[1.        , 0.5       ],
       [0.33333333, 0.25      ]])

____________________________________________![image.png](attachment:85fe41cf-7053-41c8-af99-29afb236745b.png)

____________________________________________![image.png](attachment:e098f400-88dd-4718-a254-9bad8f1d50a8.png)

<html> <h3 style="color:blue;"> Matrix Multiplication </h3> </html>

In [50]:
# Note that the above was elementwise multiplication, not matrix multiplication.
# To receive the matrix multiplication, we use the dot function.

#### Now just press 'CTRL' and 'ENTER' keys on the next cell.

![image.png](attachment:4775c518-47b9-4860-8f55-196766891597.png)

In [51]:
x = np.array([[1,2],[3,4]])
y = np.array([[5,6],[7,8]])

v = np.array([9,10])
w = np.array([11,12])
print(x, '\n\n', y, '\n\n', v, '\n\n' , w)


[[1 2]
 [3 4]] 

 [[5 6]
 [7 8]] 

 [ 9 10] 

 [11 12]


____________________________________________![image.png](attachment:f04b0822-51b3-4ca0-9645-e930c515bb4b.png)

![image.png](attachment:86e4646b-afd4-45a8-bca2-92656bfba0fb.png)

Watch this video carefully at **home** and remember how matrices are multiplied

https://www.youtube.com/watch?v=2spTnAiQg4M


____________________________________________![image.png](attachment:fd6fab27-ad50-469c-95af-88d37aa1565c.png)

In [52]:
print(v.dot(w))
print(np.dot(v, w))


219
219


In [53]:
# an example of matrix multiplication:

![image.png](attachment:134eb75e-46f0-4059-a113-23d0eb1e4b03.png)

____________________________________________![image.png](attachment:88ed560c-87be-4ced-bf7e-e8f7f1c9d370.png)

In [54]:
print(x.dot(y))
print('\n')
print(np.dot(x, y))


[[19 22]
 [43 50]]


[[19 22]
 [43 50]]


<html> <h3 style="color:blue;"> Random number array generator </h3> </html>

____________________________________________![image.png](attachment:b66acf21-ce8f-41c8-bda0-8446e71611f6.png)

In [55]:
np.random.rand(4)

array([0.20766468, 0.83012328, 0.7953583 , 0.78304057])

____________________________________________![image.png](attachment:d502bc7f-c406-47cf-86b0-858b69aa0feb.png)

In [56]:
np.random.rand(4,4)

array([[0.86358203, 0.50630323, 0.02342778, 0.63712782],
       [0.94219802, 0.28203151, 0.42084889, 0.37447717],
       [0.48420099, 0.3603615 , 0.98837877, 0.44390493],
       [0.74614173, 0.77167982, 0.25781959, 0.95514769]])

##### Normal distribution (Gaussian distribution)

![image.png](attachment:2a8d6d99-a743-406b-af0b-bded80612de1.png)

______________________![image.png](attachment:197e75d0-c7df-4e33-bbef-34d556c86be8.png)

In [57]:
np.random.randn(4,4)

array([[-1.35542876e+00, -1.75159420e-01, -2.09495993e-02,
        -8.56724502e-01],
       [ 1.43140691e+00, -1.43126794e+00, -9.38030803e-04,
         2.46218347e+00],
       [ 8.66159595e-01, -1.33262789e+00, -8.88414305e-01,
         1.53917862e+00],
       [-1.02554586e+00,  3.37275807e-01,  6.59903072e-01,
        -8.58495111e-01]])

____________________________________________![image.png](attachment:cd9a9f30-f9b9-42fb-8243-ddb743ee01c3.png)

In [58]:
np.random.randint(1,300)

234

____________________________________________![image.png](attachment:998bb937-89a8-416e-99da-0109fb725b6e.png)

In [59]:
np.random.randint(1,200,15)

array([ 70,  84,  25,  26, 102,  75, 161,  47,  11,  30, 137, 173,  27,
       193, 126])

<html> <h3 style="color:blue;"> Reshape </h3> </html>

The shape (dimensions) of the array can be changed using the reshape() method
This returns an array containing the same data with a new shape.

____________________________________________![image.png](attachment:288fc1d4-3930-4e62-8ec9-63be664670c3.png)

In [60]:
p = np.arange(0,15)

____________________________________________![image.png](attachment:3c35d8d2-17f4-43bc-a513-155af5c09df6.png)

In [61]:
p = p.reshape(3,5)

____________________________________________![image.png](attachment:2ca2045d-3dba-4c89-aeac-8176014a3cdf.png)

In [62]:
s = np.arange(12).reshape(3,4)

##### Transposing a Matrix

![image.png](attachment:ae787138-a47b-4140-9bec-99a532903bbd.png)

____________________________________________![image.png](attachment:1e9d411f-d355-4233-9180-639e7659446d.png)

In [63]:
print (p.T)

[[ 0  5 10]
 [ 1  6 11]
 [ 2  7 12]
 [ 3  8 13]
 [ 4  9 14]]


In [64]:
K = np.empty_like(p)

____________________________________________![image.png](attachment:f5b163a3-7ce4-4100-942b-efc07e220eed.png)

The command does not create an "empty" matrix in the sense of being filled with zeros or having no values. Instead, it returns a new array with the same shape and data type as p, but without initialising the elements — meaning the array elements contain random values that happened to be in memory at the time of allocation

In [65]:
v = np.array([1, 0, 1])
print(v,  '\n')

vv = np.tile(v, (4,1))
print(vv)

[1 0 1] 

[[1 0 1]
 [1 0 1]
 [1 0 1]
 [1 0 1]]


____________________________________________![image.png](attachment:4bc158d4-9ab4-45cb-8f92-48339dd3e6e5.png)

<html> <h3 style="color:blue;"> Functions </h3> </html>

____________________________________________![image.png](attachment:bf6cf184-57c9-4dd4-bfb2-7def73549040.png)

In [66]:
print(np.sum(a))

164


In [67]:
# Numpy provides many useful functions for performing computations on arrays; one of the most useful is sum:

____________________________________________![image.png](attachment:a7779684-6a28-49f0-9c2f-4fab2a79a47e.png)

In [68]:
print(np.sum(a, axis=0))


[ 15 104  21  24]


____________________________________________![image.png](attachment:26e7377d-c022-4e06-8ade-0f49b2759669.png)

In [69]:
print(np.sum(a, axis=1))

[96 26 42]


____________________________________________![image.png](attachment:e7c61121-ac0a-43d5-a824-f12b26e99c73.png)

In [70]:
a.max()

np.int64(88)

____________________________________________![image.png](attachment:0d80160c-983f-4bcf-9a16-6ce6911f9d4e.png)

In [71]:
a.min()

np.int64(1)

____________________________________________![image.png](attachment:16c464a1-f29f-4bd6-8de8-e86eab73c66d.png)

In [72]:
a.argmax()

np.int64(1)

____________________________________________![image.png](attachment:7ed980b3-88aa-43ee-a31d-3d664d24d492.png)

In [73]:
a.argmin()

np.int64(0)

____________________________________________![image.png](attachment:7ca34411-c78a-4d05-8f4b-64459220ad88.png)

<html> <h3 style="color:blue;"> Shape </h3> </html>

The size of a Numpy array can be retrieved using the attribute named "shape". Please note it is not a method. Attribute does not use "()" when evoked.

____________________________________________![image.png](attachment:c1e69ea3-9843-4361-87d3-39c77f8936e4.png)

In [74]:
arr = np.arange(10)
print(arr)
print(arr.shape)

[0 1 2 3 4 5 6 7 8 9]
(10,)


In [75]:
print(a)
print(a.shape)

[[ 1 88  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]]
(3, 4)


____________________________________________![image.png](attachment:53dde3eb-9adf-400d-9d8b-e94692a80ef5.png)

The shape of an array can be changed with various commands. Note that the following three commands all return a modified array, but do not change the original array:

____________________________________________![image.png](attachment:d481818b-d0c0-4c0b-8443-2369390cc1fa.png)

In [76]:
a.ravel()

array([ 1, 88,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12])

____________________________________________![image.png](attachment:e9a52f60-26dc-4230-97e6-b877ed8e8bba.png)

In [77]:
a.reshape(6,2)

array([[ 1, 88],
       [ 3,  4],
       [ 5,  6],
       [ 7,  8],
       [ 9, 10],
       [11, 12]])

____________________________________________![image.png](attachment:4ac28c80-ba5d-4357-a3e7-6fba7fc3d9f4.png)

In [78]:
print(a.shape)
print(a.T.shape)

(3, 4)
(4, 3)


Change matrix a to **2-d** array with 1 column. Notice the two sets of brackets for a 2-d array with one column.

Save it in another array.

Check the shape attribute value.

____________________________________________![image.png](attachment:a3063539-2ef1-4ae3-a392-e9baf8ab8ca3.png)

In [79]:
a3 = a.reshape(12,1)
print(a3)
print('\n')
print(a3.shape)

[[ 1]
 [88]
 [ 3]
 [ 4]
 [ 5]
 [ 6]
 [ 7]
 [ 8]
 [ 9]
 [10]
 [11]
 [12]]


(12, 1)


<html> <h3 style="font-style:italic; color:#FF0000;"> (!) From the point of view of mathematics - this is a 1-d vector.
However, in terms of Numpy, it is a 2-dimensional array.
</h3> </html>

# Lab Logbook Requirement:

<html> <h3 style="font-style:italic; color:blue;">
1) Create a vector using np.arange.

Determine the number of the vector elements using the following method: Take the last two digits from your SID. It should be from 00 to 99. If this number is 10 or more, it becomes the required number of the vector elements. If it is less than 10, add 100 to your number.

For example, if your SID is 2287467, and the last two digits are 67, which is greater than 10. The required number is 67. If your SID is 2287407, and the last two digits are 07, which is less than 10. The required number is 107.

Then,

2) Change matrix a to 2-d array with 1 row. Print the array.
You should have the two sets of brackets for a 2-d array with one row.
3) Save it in another array. Print the array.
4) Check the shape attribute value.
5) Add the code and result to your Lab Logbook
</h3> </html>

<html> <h3 style="color:red;">
NOTE: DON'T FORGET TO SAVE AND BACK UP YOUR COMPLETED JUPYTER NOTEBOOK AND LAB LOGBOOK ON GITHUB OR ONEDRIVE.
</h3> </html>

In [80]:
u = 42
v =np.arange(u)
print("Vector:")
print(v)

Vector:
[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23
 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41]


In [81]:
r =v.reshape(1, -1)
print("\n2d array:")
print(r)


2d array:
[[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23
  24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41]]


In [82]:
print("Shape:", r.shape)

Shape: (1, 42)
